main/distributed-programming.md

```diff
@@ -42,22 +55,35 @@ lockdown(); // turns the current realm into an immutable realm
 
-For historical reasons, the ECMAScript specification requires the *primordials* to be mutable (`Array.prototype.push = yourFunction` is valid ECMAScript but not recommended). SES provides to turn the current realm into an **immutable realm**, that is a realm within which, the primordials are deeply frozen. It also allows to create **Compartments**.
+For historical reasons, the ECMAScript specification requires the *primordials*
+to be mutable (`Array.prototype.push = yourFunction` is valid ECMAScript but not
+recommended). SES makes it possible to turn the current realm into an
+**immutable realm**, that is a realm within which the primordials are deeply
+frozen. It also allows programs to create **Compartments**.
 
-Compartments are "mini-realms". They have their own dedicated global object and environement, but they inherit the primordials from their parent realm.
+Compartments are "mini-realms". They have their own dedicated global object and environment, but they inherit the primordials from their parent realm.
 
 Agoric "deploy scripts" and smart contract code run in an immutable realm with compartments providing just enough authority to create useful and secure contracts (but not enough authority to do anything unintended or harmful to the participants of the smart contract)
 
 
-
-
 ### harden
 
 `harden` is a function provided by the [`@agoric/harden` package](https://www.npmjs.com/package/@agoric/harden).
 
-In a nutshell, `harden` creates objects that can be trusted. In practice, `harden` performs the deep version of [`Object.freeze`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze) (which only freezes the object's own properties).
-
-Most methods of the Agoric platform require their arguments to be `harden`ed. This is the only way they can provide the security properties they claim to provide.
+In a nutshell, `harden` creates objects that can't be changed. In practice,
+`harden` performs the deep version of
+[`Object.freeze`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze). Freeze
+only freezes the object's own properties, so harden is necessary in order to
+ensure that no further changes can be made. It is useful in distributed systems
+to distinguish objects with mutable state from objects that are
+unchanging. Objects that can change over time are extremely useful, but in order
+to interact with them, you have to send the message to where they are. If an
+object is guaranteed not to change, then copies of it can be widely shared, and
+interacted with directly.
+
+Many services provided by Agoric require their parameters to be hardened so they
```

Comment:
I think we're conflating hardening with something being a record. Even a presence can be hardened, but that doesn't turn it into a record that is passed by copy. The distinction is having functions or not, as I understand it.

Reply:
I don't think a user would call `harden()` on a presence. 

Perhaps you mean that presences are also hardened? I don't know whether that's true, but it doesn't seem related to what kinds of objects a user would call harden on and why.

What I'm trying to explain here is which kinds of objects a user would want to call harden on, and why.  Are there other reasons to call harden()?

Reply:
We do call harden when making objects that will be represented as presences all the time though. For instance, the Zoe service is hardened. I think we're misleading readers by associating harden at all with passing a record. 

It's a good question why we want to do that. @erights why do we do that?

Reply:
`harden` is not about immutability. It is about tamper-proofing an API surface. Our counter-example:

```js
function makeCounter() {
  let count = 0;
  return harden({
    incr() { return ++count; },
    decr() { return --count; },
  });
}
```

Returns a hardened counter object. Its whole purpose is to be mutated. Hardening the object is only about ensuring that the behavior of the object is according to the code of the object. If Alice and Bob both share the hardened counter object, they can communicate by incrementing, decrementing, and observing the count, but nothing else.

Without the harden, Alice could replace the counter's incr method with something else, causing Bob to see behavior that is not according to the code of the contract. This way madness lies.

Our general style is not to represent mutable state by modifying properties. Rather, as above, by mutating captured lexical variables. `harden` only locks down properties, nothing else.

## When it is about immutability

For objects that are deeply pass-by-copy according to marshal: they contain no functions other than inherited pure primordial methods. Without functions, they have no where else to store mutable state. Locking down all their properties, as `harden` does, makes them immutable. This case is important, but from `harden`'s pov, it's a special case.

Reply:
I can see the point of thinking about `harden()` and describing it in those terms when discussing objects that are shared within a vat, but I get confused when thinking about it in the context of writing a contract. Contract writers share objects as presences, so the remote users of those objects don't have any ability to modify them other than by their API. 

When we're trying to explain things to developers who are writing contracts in the Zoe framework, what role does this aspect of `harden()` have in what they need to understand?  I can see that it's a useful concept as far as their defensive programming if they define multiple objects over multiple source files, and want to convince themselves and reviewers of their code that they're following best practices, but it doesn't seem to add anything to the story about the contracts they're publishing.

Reply:
I think the high order role of `harden` is simply that only hardened objects are passable. Anything used as receiver, argument, or return value of a remote message must be hardened.

Reply:
I simplified the description of harden. Please take another look.